---
title: "Pool QC"
author: "Andrew Clugston"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: spacelab
    code_folding: hide
params:
  pool_name: "OS526_K27Ac_KO_Project_ACEH_EH005"
  ctrl_epitope: "IgG"
  sample_table: "stump_file_table.csv"
  output_dir: "/N/p/asclab/ASC-cutNrun/nextflow/data"
  R_dir: "R_functions"
  fastqc_trim1: "OS526_K27Ac_KO_2_val_1_fastqc.zip OS526_K27Ac_KO_1_val_1_fastqc.zip"
  fastqc_trim2: "OS526_K27Ac_KO_2_val_2_fastqc.zip OS526_K27Ac_KO_1_val_2_fastqc.zip"
  fastqc_filt: "OS526_K27Ac_KO_2_hg38_short_fastqc.zip OS526_K27Ac_KO_1_hg38_short_fastqc.zip"
  tsv_spike: "OS526_K27Ac_KO_2_spike.tsv OS526_K27Ac_KO_1_spike.tsv"
  npks_reps: "OS526_K27Ac_KO_2_peaks.narrowPeak OS526_K27Ac_KO_1_peaks.narrowPeak"
  bdgs_reps: "OS526_K27Ac_KO_2_treat_pileup.bdg.gz OS526_K27Ac_KO_1_treat_pileup.bdg.gz"
  bdgs_reps_ctrl: "OS526_K27Ac_KO_1_control_lambda.bdg.gz OS526_K27Ac_KO_2_control_lambda.bdg.gz"
  bpks_reps: "OS526_K27Ac_KO_2_peaks.broadPeak OS526_K27Ac_KO_1_peaks.broadPeak"
  tsvs_frip: "OS526_K27Ac_KO_2_narrowPeaks_FRiP.txt OS526_K27Ac_KO_1_narrowPeaks_FRiP.txt"
  npks_pool: "OS526_K27Ac_KO_peaks.narrowPeak"
  sums_pool: "OS526_K27Ac_KO_summits.bed"
  bdgs_pool: "OS526_K27Ac_KO_treat_pileup.bdg.gz"
  bdgs_pool_ctrl: "OS526_K27Ac_KO_control_lambda.bdg.gz"
  bpks_pool: "OS526_K27Ac_KO_peaks.broadPeak"
  gene_gtf: "Homo_sapiens.GRCh38.104.chr.tabix.gtf.gz"
  seqsizes: "hg38_seqsizes.tsv"
---

```{r setup, include=FALSE}
library(tidyverse) 
#library(ascCutNRun)
library(data.table)
#library(clugPac)
library(ggplot2)
library(magrittr)
library(scales)
library(knitr)
library(kableExtra)
#library(ascFunctions)
library(ggnewscale)
library(cowplot)
library(GenomicRanges)
library(reactable)
library(regioneR)
knitr::opts_chunk$set(echo = TRUE)

# Aliases.
mutate  <- dplyr::mutate
filter  <- dplyr::filter
arrange <- dplyr::arrange
rename  <- dplyr::rename
o_write <- TRUE

knitr::opts_chunk$set(echo = TRUE,fig.width = 10,fig.height = 6,warning = FALSE,message = FALSE)

###
##
#

#knitr::opts_knit$set(root.dir = "/N/p/asclab/ASC-cutNrun/nextflow/report_test")

#
##
###
param_list    <- lapply(params,function(x) str_split(x,pattern=" ")[[1]])

# Source R functions.
sapply(Sys.glob(file.path(param_list$R_dir,"*.R")),source) %>% invisible()
```

```{r loadData}
project_deets <- str_match(param_list$pool_name,"^([:alnum:]+)_([:alnum:]+)_([:alnum:]+)_(.+)")[2:5] %>% setNames(c("cell_line","epitope","condition","project"))
tb_samps      <- fread(param_list$sample_table) %>% as_tibble %>% mutate(local = project == project_deets['project']) %>% arrange(project,cell_line,epitope,cond)
gr_genes      <- gtf_to_genes(params$gene_gtf,cache_file = "genes.tsv")
gr_chroms     <- get_seqsizes(seqsize_file = param_list$seqsizes,as_granges = TRUE)
```

# File tables {.tabset}
## Pool
```{r listParameters,results='asis'}
dir_reps<- file.path(param_list$output_dir,"replicates")
dir_pool<- file.path(param_list$output_dir,"pooled",param_list$pool_name)
rep_alg <- c("tsv_spike")
rep_qc  <- c("fastqc_trim1","fastqc_trim2","fastqc_filt","tsvs_frip")
rep_pks <- c("npks_reps","bpks_reps","bdgs_reps","bdgs_reps_ctrl")
pool_pks<- c("npks_pool","sums_pool","bdgs_pool","bdgs_pool_ctrl","bpks_pool")
implicit_sample_dir <- function(fl_nm_in="OS526_K27Ac_KO_1_control_lambda.bdg.gz"){
  samp_nm <- str_extract(fl_nm_in,"^[:alnum:]+_[:alnum:]+_[:alnum:]+_[:digit:]+")
  file.path(dir_reps,samp_nm)
}
pool_nm   <- str_extract(param_list$pool_name,"^[:alnum:]+_[:alnum:]+_[:alnum:]+")

tb_params <- tibble(
    field  = names(param_list),
    value = param_list) %>%
  as_tibble %>%
  unnest(value) %>% 
  add_row(field="pool_name_simple",
          value=pool_nm) %>%
  add_row(field="tsv_spike_pooled",
          value = file.path(dir_pool,paste0(pool_nm,"_spike.tsv"))) %>%
  mutate(value = 
    case_when(
      field %in% rep_alg ~ file.path(implicit_sample_dir(value),"align",value),
      field %in% rep_qc ~ file.path(implicit_sample_dir(value),"qc",value),
      field %in% rep_pks ~ file.path(implicit_sample_dir(value),"peaks",value),
      field %in% pool_pks ~ file.path(dir_pool,"peaks",value),
      TRUE ~ value)) %T>%
    fwrite(file=paste0(param_list$pool_name,"_file_summary.tsv"),sep="\t",row.names=FALSE,col.names = TRUE,quote=FALSE)

tb_params %>%
  kable %>%
  kable_styling(full_width=FALSE)
```

## Sample table
```{r }
loc <- which(tb_samps$local)
tb_samps %>%
  select(-local) %>%
  kable %>%
  kable_styling(full_width=FALSE) %>%
  row_spec(c(1:nrow(tb_samps)),color='gray') %>%
  row_spec(loc,background='lightblue',color='black') %>%
  column_spec(6,bold="true")
```

# FastQC {.tabset}
## Raw {.tabset}
```{r fastqcRead}
fq_raw1 <- read_fastqc(param_list$fastqc_trim1)
fq_raw2 <- read_fastqc(param_list$fastqc_trim2)

fq_raw  <- lapply(names(fq_raw1), function(nm){
  tb1   <- fq_raw1[[nm]]
  tb2   <- fq_raw2[[nm]]
  
  if(inherits(tb1,"data.frame")){
    tb1 <- mutate(tb1,read="R1")
  }else{
    tb1 <- NULL
  }
  if(inherits(tb2,"data.frame")){
    tb2 <- mutate(tb2,read="R2")
  }else{
    tb2 <- NULL
  }
  if(!is.null(tb1) & !is.null(tb2)){
    tb <- rbind(tb1,tb2)
  }else if(!is.null(tb1)){
    tb <- tb1
  }else if(is.null(tb2)){
    tb <- tb2
  }else{
    tb <- NULL
  }
  return(as_tibble(tb))
}) %>% setNames(names(fq_raw1))
```

### Flag summary
```{r fastqcSummary,fig.height=6,fig.width=6}
tb_p  <- fq_raw$summary %>%
  mutate(name = gsub("_val_[12]$","",name)) %>%
  separate(name,into=c("cell_line","epitope","cond","rep"),sep = "_",remove = FALSE) %>%
  select(-cond) %>%
  mutate(cell_line = factor(cell_line),
         epitope = factor(epitope,levels=c("KLF5","IgG")),
         read = factor(read),
         rep = factor(rep),
         flag = factor(flag,levels=c("pass","warn","fail")),
         title = factor(title),
         idx = row_number()) %>%
  rowwise %>%
  mutate(x = ifelse(read == "R1",list(c(0,1,0)),list(c(0,1,1))),
         y = ifelse(read == "R1",list(c(1,1,0)),list(c(0,1,0)))) %>%
  ungroup %>%
  unnest(c(x,y)) %>%
  mutate(x = x + (as.integer(rep)-1),
         y = y + (as.integer(title)-1))

y_labs  <- levels(tb_p$title)
y_labs  <- setNames(c(1:length(y_labs))-0.5,y_labs)
x_labs  <- levels(tb_p$rep)
x_labs  <- setNames(c(1:length(x_labs))-0.5,x_labs)

ggplot(tb_p,aes(x=x,y=y,group=idx,fill=flag)) +
  facet_wrap(.~cell_line,
             nrow=1,
             scales = "free_x") +
  scale_x_continuous(name = "Replicate",
                     expand=c(0,0),
                     breaks=x_labs,
                     labels = names(x_labs)) +
  scale_y_reverse(expand=c(0,0),
                     breaks = y_labs,
                     labels = names(y_labs))+
  scale_fill_manual(values=c(pass="lightgreen",
                             warn="orange",
                             fail="firebrick")) +
  geom_polygon(color="black") +
  annotate(geom = "text",x=0.05,y=0.95,hjust=0,vjust=0,label="R1",size=2.5) +
  annotate(geom = "text",x=0.95,y=0.05,hjust=1,vjust=1,label="R2",size=2.5) +
  ggtitle("Raw reads") +
  theme_minimal() +
  theme(axis.title.y = element_blank(),
        legend.title = element_blank(),
        plot.background= element_rect(fill="white",color=NA))
```

### FastQ Reads
```{r tableStats}
tb_stats <- fq_raw$basic_statistics %>%
  filter(read == "R1") %>%
  mutate(name = gsub("_val_1","",name),
         measure = gsub(" ","_",gsub("%","pct_",tolower(measure)))) %>%
  pivot_wider(id_cols = name,
              names_from = measure,
              values_from = value) %>%
  select(name,total_sequences,total_bases,sequence_length,pct_gc) %>%
  mutate(total_sequences = as.double(total_sequences)) %>%
  left_join(by="name",
    select(tb_samps,name,cell_line,epitope,rep)) %>%
  arrange(cell_line,epitope) %>%
  mutate(name = factor(name,levels=select(.,name) %>% unlist %>% unique)) %>%
  select(cell_line,epitope,rep,everything())

tb_stats %>%
  mutate(total_sequences = prettyNumbers(total_sequences)) %>%
  kable %>%
  kable_styling(full_width=FALSE)
```
```{r plotFastqReads,fig.width=5,fig.height=5}
ggplot(tb_stats,aes(x=name,y=total_sequences,fill=epitope,label=paste0(prettyNumbers(total_sequences,digs = 1),"\n"))) +
  facet_wrap(.~cell_line,scales="free_x",nrow=1) +
  scale_y_continuous(labels = prettyNumbers,name="Reads",expand=expansion(mult=c(0,0.1))) +
  geom_bar(stat='identity',position = "dodge",color='black',linewidth=0.5) +
  geom_text(size=2) +
  ggtitle("FastQ read counts",subtitle= "Unfiltered") +
  theme(plot.background = element_rect(fill="white",color=NA),
        plot.subtitle = element_text(face="italic"),
        panel.background = element_blank(),
        panel.border = element_rect(fill = NA,color='black'),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(linewidth=0.5,color="gray80"),
        legend.title = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1,size=5),
        axis.ticks.x = element_blank(),
        strip.background = element_blank(),
        axis.title.x = element_blank())
```

### Per-base quality scores
```{r plotPerBaseSequenceQuality,fig.width=10,fig.height=4}
tb_p <- fq_raw$per_base_sequence_quality %>%
  mutate(base = ifelse(read == "R2",-base,base),
         name = gsub("_val_[12]$","",name))

ggplot(tb_p,aes(x=base,y=mean,color=name)) +
  facet_wrap(.~read,nrow=1,scales="free_x") +
  scale_x_continuous(name="Base (bp)",expand=c(0,0),labels=abs) + 
  scale_y_continuous(name="Quality score") +
  geom_line() +
  ggtitle("Per-base quality scores",subtitle= "Unfiltered") +
  theme(plot.background = element_rect(fill="white",color=NA),
        plot.subtitle = element_text(face="italic"),
        panel.background = element_blank(),
        panel.border = element_rect(fill = NA,color='black'),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(linewidth=0.5,color="gray80"),
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        strip.background = element_blank())
```

### Per-base sequence content
```{r perBaseSequenceContent,fig.height=6,fig.width=8}
tb_p <- fq_raw$per_base_sequence_content %>%
  mutate(name = gsub("_val_[12]$","",name),
         base = ifelse(read == "R2",-base,base)) %>%
  as_tibble %>%
  left_join(select(tb_samps,name,cell_line,epitope,rep),by="name") %>%
  mutate(cond = paste(cell_line,rep)) %>%
  rename(position = base) %>%
  pivot_longer(cols=c(a,c,g,t),names_to = "base",values_to="percent")

ggplot(tb_p,aes(x=position,y=percent,color=cond,linetype=base)) +
  facet_grid(rows=vars(epitope),cols=vars(read),scales="free_x") +
  scale_x_continuous(name="Base",expand=c(0,0),labels=abs) +
  scale_y_continuous(name="Percent",labels = function(x) paste0(x,"%")) +
  geom_line() +
  ggtitle("Per-base sequence content",subtitle="Unfiltered") +
  theme(plot.background = element_rect(fill="white",color=NA),
          plot.subtitle = element_text(face="italic"),
          panel.background = element_blank(),
          panel.border = element_rect(fill = NA,color='black'),
          panel.grid.major.x = element_blank(),
          panel.grid.major.y = element_line(linewidth=0.5,color="gray80"),
          legend.title = element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank())
```

### Per-sequence GC content
```{r perSequenceGCContent,fig.height=6,fig.width=8}
tb_p  <- fq_raw$per_sequence_gc_content %>%
  mutate(name = gsub("_val_[12]$","",name),
         read = factor(read)) %>% 
  left_join(select(tb_samps,name,cell_line,epitope,rep),by="name") %>%
  mutate(name = paste(cell_line,rep))

ggplot(tb_p,aes(x=gc_content,y=count,color=name,group=name)) +
  facet_grid(rows=vars(epitope),cols=vars(read),scales="free") +
  scale_x_continuous(name = "GC content",labels=function(x) paste0(x,"%")) +
  scale_y_continuous(name = "Reads",labels=prettyNumbers) +
  geom_line() +
  ggtitle("Per-sequence GC content",subtitle="Unfiltered") +
  theme(plot.background = element_rect(fill="white",color=NA),
            plot.subtitle = element_text(face="italic"),
            panel.background = element_blank(),
            panel.border = element_rect(fill = NA,color='black'),
            panel.grid.major.x = element_blank(),
            panel.grid.major.y = element_line(linewidth=0.5,color="gray80"),
            legend.title = element_blank(),
            axis.ticks.x = element_blank(),
            strip.background = element_blank())
```

## Aligned & filtered {.tabset}
```{r fastqcReadFilt}
fq_hg38 <- read_fastqc(param_list$fastqc_filt)
```

### Flag summary
```{r fastqcSummaryFilt,fig.height=6,fig.width=6}
tb_p  <- fq_hg38$summary %>%
  mutate(name = gsub("_hg38_short$","",name)) %>%
  separate(name,into=c("cell_line","epitope","cond","rep"),sep = "_",remove = FALSE) %>%
  select(-cond) %>%
  mutate(cell_line = factor(cell_line),
         epitope = factor(epitope,levels=c("KLF5","IgG")),
         rep = factor(rep),
         flag = factor(flag,levels=c("pass","warn","fail")),
         title = factor(title,levels=select(.,title) %>% unlist %>% unique %>% sort %>% rev))

ggplot(tb_p,aes(x=rep,y=title,fill=flag)) +
  facet_wrap(.~cell_line,nrow=1) +
  scale_x_discrete(name="Replicates") +
  scale_fill_manual(values=c(pass="lightgreen",
                             warn="orange",
                             fail="firebrick")) +
  geom_tile(color="black") +
  ggtitle("Filtered reads") +
  theme_minimal() +
  theme(axis.title.y = element_blank(),
        legend.title = element_blank(),
        plot.background= element_rect(fill="white",color=NA),
        panel.grid = element_blank())
```

### FastQ Reads
```{r tableStatsFilt}
tb_stats <- fq_hg38$basic_statistics %>%
  as_tibble %>%
  mutate(name = gsub("_hg38_short","",name),
         measure = gsub(" ","_",gsub("%","pct_",tolower(measure)))) %>%
  pivot_wider(id_cols = name,
              names_from = measure,
              values_from = value) %>%
  select(name,total_sequences,total_bases,sequence_length,pct_gc) %>%
  mutate(total_sequences = as.double(total_sequences)) %>%
  left_join(by="name",
    select(tb_samps,name,cell_line,epitope,rep)) %>%
  arrange(cell_line,epitope) %>%
  mutate(name = factor(name,levels=select(.,name) %>% unlist %>% unique)) %>%
  select(cell_line,epitope,rep,everything())

tb_stats %>%
  mutate(total_sequences = prettyNumbers(total_sequences)) %>%
  kable %>%
  kable_styling(full_width=FALSE)
```

```{r plotFastqReadsFilt,fig.height=4,fig.width=8}
ggplot(tb_stats,aes(x=name,y=total_sequences,fill=epitope,label=paste0(prettyNumbers(total_sequences,digs = 1),"\n"))) +
  facet_wrap(.~cell_line,scales="free_x",nrow=1) +
  scale_y_continuous(labels = prettyNumbers,name="Reads",expand=expansion(mult=c(0,0.1))) +
  #scale_fill_manual(values=c(KLF5="lightblue",IgG = "gray50")) +
  geom_bar(stat='identity',position = "dodge",color='black',linewidth=0.5) +
  geom_text(size=2) +
  ggtitle("FastQ read counts",subtitle= "Filtered") +
  theme(plot.background = element_rect(fill="white",color=NA),
        plot.subtitle = element_text(face="italic"),
        panel.background = element_blank(),
        panel.border = element_rect(fill = NA,color='black'),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(linewidth=0.5,color="gray80"),
        legend.title = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1,size=5),
        axis.ticks.x = element_blank(),
        strip.background = element_blank(),
        axis.title.x = element_blank())
```

### Per-base quality scores
```{r plotPerBaseSequenceQualityFilt,fig.width=8,fig.height=4}
tb_p <- fq_hg38$per_base_sequence_quality %>%
  mutate(name = gsub("_hg38_short$","",name))

ggplot(tb_p,aes(x=base,y=mean,color=name)) +
  scale_x_continuous(name="Base (bp)",expand=c(0,0),labels=abs) + 
  scale_y_continuous(name="Quality score") +
  geom_line() +
  ggtitle("Per-base quality scores",subtitle= "Filtered") +
  theme(plot.background = element_rect(fill="white",color=NA),
        plot.subtitle = element_text(face="italic"),
        panel.background = element_blank(),
        panel.border = element_rect(fill = NA,color='black'),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(linewidth=0.5,color="gray80"),
        legend.title = element_blank(),
        axis.ticks.x = element_blank(),
        strip.background = element_blank())
```


### Per-base sequence content
```{r perBaseSequenceContentFilt,fig.width=8,fig.height=5}
tb_p <- fq_hg38$per_base_sequence_content %>%
  mutate(name = gsub("_hg38_short$","",name)) %>%
  as_tibble %>%
  left_join(select(tb_samps,name,cell_line,epitope,rep),by="name") %>%
  mutate(cond = paste(cell_line,rep)) %>%
  rename(position = base) %>%
  pivot_longer(cols=c(a,c,g,t),names_to = "base",values_to="percent")

ggplot(tb_p,aes(x=position,y=percent,color=cond,linetype=base)) +
  facet_grid(rows=vars(epitope)) +
  scale_x_continuous(name="Base",expand=c(0,0),labels=abs) +
  scale_y_continuous(name="Percent",labels = function(x) paste0(x,"%")) +
  geom_line() +
  ggtitle("Per-base sequence content",subtitle="Filtered") +
  theme(plot.background = element_rect(fill="white",color=NA),
          plot.subtitle = element_text(face="italic"),
          panel.background = element_blank(),
          panel.border = element_rect(fill = NA,color='black'),
          panel.grid.major.x = element_blank(),
          panel.grid.major.y = element_line(linewidth=0.5,color="gray80"),
          legend.title = element_blank(),
          axis.ticks.x = element_blank(),
          strip.background = element_blank())
```

### Per-sequence GC content
```{r perSequenceGCContentFilt,fig.width=8,fig.height=5}
tb_p  <- fq_hg38$per_sequence_gc_content %>%
  mutate(name = gsub("_hg38_short$","",name)) %>% 
  left_join(select(tb_samps,name,cell_line,epitope,rep),by="name") %>%
  mutate(name = paste(cell_line,rep))

ggplot(tb_p,aes(x=gc_content,y=count,color=name,group=name)) +
  facet_grid(rows=vars(epitope)) +
  scale_x_continuous(name = "GC content",labels=function(x) paste0(x,"%")) +
  scale_y_continuous(name = "Reads",labels=prettyNumbers) +
  geom_line() +
  ggtitle("Per-sequence GC content",subtitle="Filtered") +
  theme(plot.background = element_rect(fill="white",color=NA),
            plot.subtitle = element_text(face="italic"),
            panel.background = element_blank(),
            panel.border = element_rect(fill = NA,color='black'),
            panel.grid.major.x = element_blank(),
            panel.grid.major.y = element_line(linewidth=0.5,color="gray80"),
            legend.title = element_blank(),
            axis.ticks.x = element_blank(),
            strip.background = element_blank())
```

# Alignment

Alignment percentages are calculated as **reads successfully mapped to hg38 or sac3 genomes / total reads**.

When calculating scores between samples, the ratio of *Sac3* to *hg38* reads can be used to normalize: we assume all samples had an equivalent number of *Sac3* reads spiked, and thus **any missing *Sac3* reads indicate missing *hg38* reads as well**. For this reason, we can *scale down reads of all samples to match the sample with the lowest ratio*. In other words, **we assume the lowest ratio sample lost reads and so we down-sample all others to match it.**

For instance, if **SampleA** has a ratio of 0.02 and **SampleB** has a ratio of 0.01, each ratio is divided into 0.01 to give factors of 0.5 (**SampleA**) and 1 (**SampleB**). A peak with 100 reads in **SampleA** and 50 reads in **SampleB** would be multiplied by their respective factors to both be 50.
```{r loadSpikes,fig.height=6,fig.width=8}
type_levs=c(ambiguous="gray",unmapped="pink",sac3="purple",hg38="darkgreen")
tb_p <- lapply(param_list$tsv_spike,function(fl_nm) {
  fread(fl_nm,col.names=c("type","reads")) %>%
    mutate(type = 
      case_when(grepl("_hg38_",type) ~ "hg38",
                grepl("_sac3_",type) ~ "sac3",
                TRUE ~ type),
           name = gsub("_spike.tsv","",fl_nm))
  }) %>% do.call(rbind,.) %>%
  group_by(name) %>%
  mutate(type = factor(type,levels=names(type_levs))) %>%
  ungroup

tb_labs <- tb_p %>%
  mutate(stat = ifelse(type %in% c("hg38","sac3"),"aligned","unaligned")) %>%
  group_by(name,stat) %>%
  summarize(total = sum(reads),.groups='drop_last') %>%
  mutate(frac = total / sum(total)) %>%
  ungroup %>%
  filter(stat=='aligned')

p1 <- ggplot(tb_p,aes(x=name,y=reads,fill=type)) +
  scale_y_continuous(name = "Reads",labels=prettyNumbers,expand=expansion(mult=c(0,0.1))) +
  scale_fill_manual(values=type_levs) +
  geom_bar(stat='identity') +
  geom_text(data=tb_labs,mapping=aes(x=name,y=total,label=paste0("\n",percent(frac,accuracy=1))),
            color="white",inherit.aes=FALSE) +
  ggtitle("Read alignment",subtitle="Alignment percentage calculated from hg38 + sac3 reads.") +
  theme(plot.background = element_rect(fill="white",color=NA),
        plot.subtitle = element_text(face="italic"),
        panel.background = element_blank(),
        panel.border = element_rect(fill=NA,color="black"),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(linewidth=0.5,color='gray75'),
        strip.background = element_blank(),
        axis.text.x = element_text(angle=90,hjust=1,vjust=0.5),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.title = element_blank())

tb_p2 <- tb_p %>%
  filter(type %in% c("sac3","hg38")) %>%
  group_by(name) %>%
  mutate(ratio = reads / sum(reads)) %>%
  filter(type == "sac3")
p2 <- ggplot(tb_p2,aes(x=name,y=ratio,label=paste0("\n",percent(ratio,accuracy=0.01)))) +
  scale_y_reverse(name= "Sac3/hg38 Ratio",labels=percent,expand=expansion(mult=c(0.1,0))) +
  geom_bar(stat='identity',fill='purple',color="black",linewidth=0.25) +
  geom_text() +
  theme(plot.background = element_rect(fill="white",color=NA),
        plot.subtitle = element_text(face="italic"),
        panel.background = element_blank(),
        panel.border = element_rect(fill=NA,color="black"),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(linewidth=0.5,color='gray75'),
        strip.background = element_blank(),
        strip.text = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        legend.title = element_blank())

plot_grid(p1,p2,ncol=1,align='v',axis='lr',rel_heights = c(5,3))
```

# Peak calling {.tabset}

For individual samples, peaks are called using the genome-wide median as a background.

## Enrichment {.tabset}

Assuming that measured peaks are real, we expect them to fall on or near promoters (within 10kb of a TSS) or enhancers (within 500kb of a TSS). In the following plots, the distance between each peak and its nearest TSS is calculated and binned. To determine if these values are *significant*, we compare them to a *random distribution of the same peaks*. For every peak, we randomize its location (but not size) within the genome 20 times and measure its distance from a TSS each time. The red segments/curves thus show a normal distribution: i.e., how far do most peaks fall from a TSS if they are distributed randomly. Fold enrichment of measured data is then calculated as the ratio of **real values / randomized values**, and labeled on top of each bar. A value of 2x, for instance, indicates that peaks are *twice as likely to fall in a given bin than expected by chance*, while a value of 1x indicates that we would expect the same result if peaks were random, i.e. not enriched. In most cases, if no bins are substantially greater than 1, we may expect the dataset to be problematic.
```{r }
tally_nearest <- function(peaks_in,regions_in,brks,brk_nms){
  distanceToNearest(peaks_in,regions_in) %>%
    as_tibble %>%
    left_join(
      as_tibble(peaks_in) %>% select(seqnames,start,end) %>% rename_all(function(x) paste0("peak_",x)) %>% mutate(queryHits=row_number()),
      by="queryHits"
    ) %>%
    left_join(
      as_tibble(regions_in) %>% select(seqnames,start,end,strand) %>% rename_all(function(x) paste0("gene_",x)) %>% mutate(subjectHits=row_number()),
      by="subjectHits"
    ) %>%
    mutate(distance = case_when(gene_strand == "+" & peak_end < gene_start ~ -distance,
                                gene_strand == "-" & peak_start > gene_end ~ -distance,
                                TRUE ~ distance),
           bin = cut(distance,breaks = brks,include.lowest = TRUE,labels = brk_nms)) %>%
    group_by(bin) %>%
    tally(name='count') %>%
    ungroup %>%
    mutate(frac = count/sum(count)) %>%
    return()
}
bootstrap_enrichment <- function(peaks_in,regions_in,brks_in,brk_nms_in,iters=10,genome_in=gr_chroms){
  gr_rand <- randomize_grange(peaks_in,max_sample = Inf,sample_num = iters,genome_in = genome_in)
  
  rbind(
    tally_nearest(peaks_in,regions_in,brks = brks_in,brk_nms=brk_nms_in) %>% mutate(name = "real"),
    tally_nearest(gr_rand,regions_in,brks = brks_in, brk_nms=brk_nms_in) %>% mutate(name = "rand",count=count/iters)
  ) %>%
    pivot_wider(id_cols = bin,names_from = name,values_from = c(count,frac),values_fill = 0) %>%
    mutate(ratio_frac = frac_real/frac_rand,
           ratio_count = count_real /  count_rand,
           name = peaks_in$name[1]) %>%
    return()
}
plot_peak_enrichment <- function(peak_fls,chrom_gr=gr_chroms,iterations=20){
 brks <- c(-Inf,-1E6,-5E5,-1E5,-5E4,-1E4,1E4,5E4,1E5,5E5,1E6,Inf)
brk_nms <- c(
  ">1MB",
  "1MB-500kb",
  "500kb-100kb",
  "100kb-50kb",
  "50kb-10kb",
  "TSS +/-10kb",
  "10kb-50kb",
  "50kb-100kb",
  "100kb-500kb",
  "500kb-1MB",
  ">1MB") %>%
  paste0(c(rep(" upstream",5),"",rep(" downstream",5)))
  
  tb_enr<- lapply(names(peak_fls),function(nm) {
      read_peak_file(peak_fls[[nm]]) %>%
        filter(seqnames %in% seqlevels(chrom_gr)) %>%
        makeGRangesFromDataFrame %>%
        reduce %>%
        bootstrap_enrichment(
          regions_in = promoters(gr_genes,upstream=1,downstream=0),
          brks_in = brks,
          brk_nms_in = brk_nms,
          genome_in = chrom_gr,
          iters=iterations) %>%
        mutate(name = nm)
    }) %>% do.call(rbind,.)
    
  ggplot(tb_enr,aes(x=bin,y=count_real,label=paste0(round(ratio_frac,digits=1),"x\n"))) +
    facet_wrap(.~name,ncol=1,scales="free_y",strip.position = "right") +
    scale_x_discrete(labels = function(x) gsub(" ","\n",x)) +
    scale_y_continuous(name="Peaks",expand=expansion(mult=c(0,0.2)),labels=prettyNumbers) +
    geom_ribbon(mapping=aes(x=bin,ymax=count_rand,ymin=0,group="group"),color="red",fill="black",alpha=0.2) +
    geom_ribbon(mapping=aes(x=bin,ymax=count_real,ymin=0,group="group"),color="black",fill="black",alpha=0) +
    geom_bar(stat='identity',color="black") +
    geom_segment(mapping=aes(x=as.integer(bin)-0.15,xend=as.integer(bin)+0.15,
                             y=count_rand,yend=count_rand),
                 color='red') +
    geom_text(size=2) +
    ggtitle("Distance to nearest TSS",subtitle=paste("Mean among",iterations,"randomized iterations in red.")) +
    theme(plot.background = element_rect(fill="white",color=NA),
          plot.subtitle = element_text(face="italic"),
          panel.background = element_blank(),
          panel.border = element_rect(fill=NA,color="black",linewidth=0.5),
          panel.grid = element_blank(),
          axis.text.x = element_text(size=4,angle=45,hjust=1,vjust=1),
          axis.ticks.x = element_blank(),
          axis.title.x = element_blank(),
          strip.background = element_blank(),
          strip.text.y.right = element_text(angle=0,hjust=0,size=6))
}
```

### narrowPeaks
```{r fig.height=6,fig.width=8}
c(param_list$npks_pool,param_list$npks_reps) %>%
  setNames(gsub("_peaks.narrowPeak","",.)) %>%
  plot_peak_enrichment
```

### broadPeaks
```{r fig.height=6,fig.width=8}
c(param_list$bpks_pool,param_list$bpks_reps) %>% 
  setNames(gsub("_peaks.broadPeak","",.)) %>%
  plot_peak_enrichment
```

## FRiP
```{r indPeakcalling,height=4,width=6}
tb_p <- lapply(param_list$tsvs_frip,function(fl_nm) {
  fread(fl_nm) %>%
    as_tibble %>%
    mutate(name = gsub("_narrowPeaks_FRiP.txt","",fl_nm))
}) %>% do.call(rbind,.)

tb_lab <- tibble(x=0.5,y=0.3,label="Ideal FRiP\n(Cut&Run)",cell_line=tb_p$cell_line[1])

ggplot(tb_p,aes(x=name,y=frip)) +
  scale_y_continuous(name="FRiP",labels=percent,limits=c(0,1),expand=c(0,0)) +
  geom_bar(stat='identity',color='black',linewidth=0.25) +
  geom_hline(yintercept = 0.3,color='red',linetype='dashed') +
  geom_text(data=tb_lab,mapping=aes(x=x,y=y,label=label),
            color="red",hjust=0,
            inherit.aes=FALSE) +
  ggtitle("Fraction of Reads in Peaks") +
  theme(plot.background = element_blank(),
        panel.background = element_blank(),
        panel.border = element_rect(fill=NA,color="black",linewidth = 0.5),
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_line(linewidth=0.5,color='gray85'),
        strip.background = element_blank(),
        axis.text.x = element_text(angle=45,hjust=1,vjust=1),
        axis.title.x = element_blank())
```

## Peak stats {.tabset}
```{r peakFunction}
peak_plot <- function(peak_file_list,p_title="NarrowPeaks",file_suffix="_peaks.narrowPeak",theme_addendum=theme()){
  tb_pks  <- lapply(peak_file_list,function(fl_nm){
  read_peak_file(fl_nm) %>%
    as_tibble %>%
    mutate(name = gsub(file_suffix,"",fl_nm))
  }) %>% do.call(rbind,.) %>%
    mutate(width = end - start)

  tb_p <- tb_pks %>%
    group_by(name) %>%
    tally(name = "count") %>%
    ungroup
  
  tm_vio <- theme(
    plot.background = element_rect(fill="white",color=NA),
    panel.background = element_blank(),
    panel.border = element_rect(fill=NA,color="black"),
    panel.grid.major.x = element_blank(),
    panel.grid.major.y = element_line(linewidth=0.5,color='gray85'),
    strip.background = element_blank(),
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank(),
    legend.title = element_blank(),
    legend.position = 'none') +
    theme_addendum
  
  p_c <- ggplot(tb_p,aes(x=name,y=count)) +
    scale_y_continuous(expand=expansion(mult=c(0,0.1)),labels=prettyNumbers,name="Peaks") +
    geom_bar(stat="identity",color='black',linewidth=0.5) +
    ggtitle(p_title) +
    tm_vio
    
  
  p_q <- ggplot(tb_pks,aes(x=name,y=nlog10q)) +
    scale_y_log10(name="-log10(Q)",labels = prettyNumbers) +
    geom_violin(draw_quantiles = 0.5,scale = "width",fill="gray85") +
    tm_vio
  
  p_w <- ggplot(tb_pks,aes(x=name,y=width,fill=epitope)) +
    scale_y_log10(name="Width",labels = prettyBP) +
    geom_violin(draw_quantiles = 0.5,scale = "width",fill="gray85") +
    tm_vio +
    theme(axis.text.x = element_text(angle=45,hjust=1,vjust=1))
  
  cowplot::plot_grid(p_c,p_q,p_w,ncol=1,align='v',axis='lr',rel_heights = c(2,2,3))
}
```

### Replicates
```{r peakDetailsRepsFig.height=8,fig.width=6}
p_nar <- peak_plot(param_list$npks_reps,"NarrowPeaks","_peaks.narrowPeak")
p_brd <- peak_plot(param_list$bpks_reps,"BroadPeaks","_peaks.broadPeak",theme_addendum = theme(axis.title.y = element_blank()))

cowplot::plot_grid(p_nar,p_brd,nrow=1,align='v',axis='tb')
```

### Pooled
```{r peakDetailsPooledFig.height=8,fig.width=6}
p_nar <- peak_plot(param_list$npks_pool,"NarrowPeaks","_peaks.narrowPeak")
p_brd <- peak_plot(param_list$bpks_pool,"BroadPeaks","_peaks.broadPeak",theme_addendum = theme(axis.title.y = element_blank()))

cowplot::plot_grid(p_nar,p_brd,nrow=1,align='v',axis='tb')
```

# Regional plots {.tabset}
```{r regionPlotFunctions}
read_peaks <- function(peak_file_list){
  lapply(peak_file_list, function(fl_nm) {
    read_peak_file(fl_nm) %>%
      mutate(seqnames = factor(seqnames,levels=paste0("chr",c(1:22,"X","Y")))) %>%
      filter(!is.na(seqnames)) %>%
      makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>%
      annotate_and_reduce_gr(retain_funcs = 
        list(name=function(x) x[1],
             score = max,
             fc_summit = max,
             nlog10p = max,
             nlog10q = max)) %>%
      as_tibble
    }) %>% do.call(rbind,.)
}
plot_region <- function(gr_window,bdg_tbl = tb_bdgs,gr_peaks_in = gr_pks){
  x_rng   <- c(start(gr_window),end(gr_window))
  tb_sig  <- bdg_tbl %>%
    vectify(bdg,name) %>%
    scan_bdg(gr_regions = gr_window) %>%
    left_join(select(bdg_tbl,-bdg),by="name") %>%
    group_by(name) %>%
    mutate(x = case_when(start == min(start) ~ start,
                         end == max(end) ~ end,
                         TRUE ~ (start + end) / 2),
           score = ifelse(type == "ctrl",-score,score))
  
  tb_scales <- tb_sig %>%
    group_by(samp) %>%
    summarize(ymin=0,ymax=1.3*max(score),.groups='drop')
  
  tb_pks <- subset_and_trim_gr(gr_peaks_in,gr_window) %>% 
    as_tibble %>%
    rename(samp = name) %>%
    left_join(tb_scales,by="samp") %>%
    mutate(y_shift = 0.05 * (ymax - ymin),
           y1=case_when(peaks == "broadPeaks" ~ ymax,
                        peaks == "narrowPeaks" ~ ymax - y_shift,
                        TRUE ~ ymax - 3*y_shift),
           y2=y1 - y_shift)
  tb_s  <- filter(tb_pks,peaks=="summits")
  tb_p  <- filter(tb_pks,peaks!="summits")
  
  geom_pks <- NULL
  geom_sums <- NULL
  if(nrow(tb_s) > 0){
    geom_sums <- geom_point(data=tb_s,mapping=aes(x=(start+end)/2,y=y1),color="black",pch=8,inherit.aes=FALSE)
  }
  if(nrow(tb_p) > 0){
    geom_pks<- geom_rect(data=tb_p,mapping=aes(xmin=start,xmax=end,ymin=y2,ymax=y1,fill=peaks),color='black',alpha=1,linewidth=0.1,inherit.aes=FALSE)
  }
  p_sig <- ggplot(tb_sig,aes(x=x,ymin=0,ymax=score,alpha=type)) +
    facet_wrap(.~samp,scales="free_y",strip.position = "right",ncol=1) +
    scale_x_continuous(name= grange_desc(gr_window),expand=c(0,0),limits=x_rng,labels=prettyBP) +
    scale_y_continuous(name= "Pileup",labels=function(x) ifelse(x>0,prettyNumbers(x),""),n.breaks = 3) +
    scale_alpha_manual(values=c(treat=1,ctrl=0.3)) +
    geom_ribbon(fill="gray25",color='black') +
    geom_pks +
    geom_sums +
    new_scale_fill() +
    geom_rect(data=tb_scales,mapping=aes(ymax=ymax),xmin=-Inf,xmax=Inf,ymin=0,fill=NA,color=NA,inherit.aes=FALSE) +
    theme(plot.background = element_rect(fill="white",color=NA),
          panel.background = element_blank(),
          panel.border = element_rect(fill=NA,color="black"),
          panel.grid = element_blank(),
          panel.spacing.y = unit(0,"lines"),
          strip.text.y.right = element_text(angle=0),
          strip.background = element_blank(),
          legend.title = element_blank()) %>% suppressWarnings()
  p_gns <- plot_genes(genomic_region = gr_window,gr_genes = gr_gns) +
    theme(legend.position = 'none',
          panel.grid = element_blank())
  
  cowplot::plot_grid(p_gns,p_sig,ncol=1,align="v",axis='lr',rel_heights=c(1,5))
}
```
```{r }
gr_gns <- gtf_to_genes(params$gene_gtf)
gr_pks <- rbind(
  read_peaks(param_list$npks_reps) %>% mutate(peaks="narrowPeaks",set="reps"),
  read_peaks(param_list$bpks_reps) %>% mutate(peaks="broadPeaks",set="reps"),
  read_peaks(param_list$npks_pool) %>% mutate(peaks="narrowPeaks",set="pool"),
  read_peaks(param_list$bpks_pool) %>% mutate(peaks="broadPeaks",set="pool")) %>%
  mutate(name = gsub("_all_peak_.+$","",name)) %>%
  select(seqnames,start,end,name,fc_summit,peaks,set) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

gr_sums <- fread(param_list$sums_pool,col.names=c("seqnames","start","end","name","fc_summit")) %>% 
  mutate(peaks="summits",
         set="pool",
         name = gsub("_all_peak_.+$","",name)) %>%
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

gr_pks <- c(gr_pks,gr_sums)

tb_bdgs <- list(
  tibble(param_list$bdgs_reps,"reps","treat"),
  tibble(param_list$bdgs_reps_ctrl,"reps","ctrl"),
  tibble(param_list$bdgs_pool,"pool","treat"),
  tibble(param_list$bdgs_pool_ctrl,"pool","ctrl")
  ) %>% lapply(function(tb) {
    colnames(tb) <- c("bdg","set","type")
    return(tb)
  }) %>% do.call(rbind,.) %>%
  mutate(set = factor(set,levels=c("pool","reps")),
         type = factor(type,levels=c("treat","ctrl")),
         name = ifelse(type == "ctrl",
                       gsub("_lambda.bdg.gz","",bdg),
                       gsub("_pileup.bdg.gz","",bdg)),
         samp = gsub("_[^_]+$","",name))

gr_npks   <- read_peaks(param_list$npks_pool) %>% makeGRangesFromDataFrame(keep.extra.columns = TRUE)
pk_samples<- min(c(length(gr_npks),5))

gr_wins <- gr_npks[order(gr_npks$nlog10q,decreasing = TRUE)][1:pk_samples] %>% resize(width=width(.) * 10,fix='center')

for(i in 1:pk_samples){
  plot(plot_region(gr_window = gr_wins[i]))
}
```